porch/repository/pkg/git/draft.go

```diff
@@ -27,111 +26,171 @@ import (
+		if err != nil {
+			return nil, fmt.Errorf("failed to finalize package: %v", err)
+		}
+
+		pushRefSpecs = append(
+			pushRefSpecs,
+			config.RefSpec(fmt.Sprintf("%s:%s", commitHash, remoteRef)),      // Push new main
+			config.RefSpec(fmt.Sprintf("%s:%s", commitHash, refNames.final)), // Push the tag
+		)
+
+		if d.ref != nil {
+			cleanup[d.ref.Name()] = true
+		}
+		// also make sure to clean up draft and proposed refs for this package, if they exist.
+		cleanup[refNames.draft] = true
+		cleanup[refNames.proposed] = true
+
+		// Update package draft
+		d.commit = commitHash
+		d.tree = newTreeHash
+		d.ref = newRef
+
+	case v1alpha1.PackageRevisionLifecycleProposed:
+		newRef, err := setReference(repo.Storer, refNames.proposed, d.commit)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update package lifecycle to \"proposed\": %v", err)
+		}
+		// Push the package revision into a proposed branch.
+		pushRefSpecs = append(pushRefSpecs, config.RefSpec(fmt.Sprintf("%s:%s", d.commit, refNames.proposed)))
+		cleanup[refNames.draft] = true
+
+		// Update package referemce (commit and tree hash stay the same)
+		d.ref = newRef
+
+	case v1alpha1.PackageRevisionLifecycleDraft:
+		newRef, err := setReference(repo.Storer, refNames.draft, d.commit)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update package draft: %v", err)
+		}
+		// Push the package revision into a draft branch.
+		pushRefSpecs = append(pushRefSpecs, config.RefSpec(fmt.Sprintf("%s:%s", d.commit, refNames.draft)))
+		cleanup[refNames.proposed] = true // In case client downgraded packaget to draft from proposed
+
+		// Update package referemce (commit and tree hash stay the same)
+		d.ref = newRef
+	}
 
 	auth, err := d.parent.getAuthMethod(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to obtain git credentials: %w", err)
 	}
 
 	if err := d.parent.repo.Push(&git.PushOptions{
-		RemoteName: "origin",
-		RefSpecs:   []config.RefSpec{refSpec},
+		RemoteName: originName,
+		RefSpecs:   pushRefSpecs,
 		Auth:       auth,
 		Force:      true, // TODO: implement conflict recovery.
 	}); err != nil {
 		return nil, fmt.Errorf("failed to push to git: %w", err)
 	}
 
+	deleteRemotes := []config.RefSpec{}
+	// Cleanup local and remote branches
+	for rn := range cleanup {
+		switch err := d.parent.repo.Storer.RemoveReference(rn); err {
+		case nil, plumbing.ErrReferenceNotFound:
+			// These are OK.
+
+		default:
+			return nil, err
+		}
+
+		// Check if its corresponding remote exists locally
+		if strings.HasPrefix(rn.String(), refHeadsPrefix) {
+			remote := plumbing.NewRemoteReferenceName("origin", strings.TrimPrefix(rn.String(), refHeadsPrefix))
+			if _, err := d.parent.repo.Reference(remote, true); err == nil {
+				deleteRemotes = append(deleteRemotes, config.RefSpec(fmt.Sprintf(":%s", rn.String())))
+			}
+		}
+	}
+
+	if len(deleteRemotes) > 0 {
+		if err := d.parent.repo.Push(&git.PushOptions{
+			RemoteName: "origin",
```

Comment:
Should this be originName?

Reply:
Oh, it's a const, so I guess it doesn't much matter :-)

Reply:
it should. Fixed. Origin is unlikely to change (because Porch is in control of it) but branch names we'll need to make generic/configurable so it is a good hygiene to reduce number of places to update later.